tests/test_grammar.py

```diff
@@ -137,6 +137,23 @@
     ("str_quoted_too_many_1", "''a'", GrammarParseError),
     ("str_quoted_too_many_2", "'a''", GrammarParseError),
     ("str_quoted_too_many_3", "''a''", GrammarParseError),
+    ("str_quoted_trailing_esc_1", "'abc\\\\'", "abc\\"),
```

Comment:
This seems like the obvious first test for trailing backslash. do you have it somewhere else?
```
    ("str_quoted_trailing_backslash", "'abc\\'", "abc\"),
```

Reply:
Nope, I realized such tests were missing when working on #621 (I agree this one should have been there earlier, and some of the other ones as well)

Reply:
Can you add the missing tests here?
They will make it easier for me to form a mental model of what the behavior is after this change.
some input strings single quoted strings.
```
''
'a' 
'\a'
'a\'
'"'
'\''
'c:\\''
```

Here are some nested quoted strings cases, some may be handled in #621:
```
'${f:"b"}' # nested double quotes
'${f:\'b\'}' # nested escaped single quotes
```
Same as above, but for an escaped interpolation. what is the expected behavior here?
enerally speaking, is parsing inside interpolations the same as parsing outside?
```
'\${f:"b"}'
'\${f:\'b\'}'
```

Reply:
> Can you add the missing tests here?

Oh, I now realize that I had probably misunderstood your previous question "This seems like the obvious first test for trailing backslash. do you have it somewhere else?" and as a result my answer was a bit off.

The example you gave:
```python
("str_quoted_trailing_backslash", "'abc\\'", "abc\"),
```
is actually invalid because `"'abc\\'"` is the string with characters `'abc\'` and is an invalid quoted string according to the grammar, since the end quote is escaped (=> it's missing another end quote).
The test for a trailing backslash is the one on the line you highlighted:
```python
("str_quoted_trailing_esc_1", "'abc\\\\'", "abc\\"),
```
where `"'abc\\\\'"` is the string with characters `'abc\\'` which is now a valid quoted string.

> (some examples)

I added all your examples in 7f23c1b
(some are named with "tmp" because they are redundant with existing tests, I still kept them here to make it easier for you to check them out, but I will remove them afterwards)
All of them behave as expected -- or at least how I would expect :)

> Generally speaking, is parsing inside interpolations the same as parsing outside?

Depends on what you mean. Generally not, because outside of interpolations we do nothing except un-escaping (because of escaped interpolations) and otherwise just copy any character that we see (there are no constraints). But when we are inside an interpolation (i.e., parsing resolver arguments), the logic is different (to detect dicts, lists, quoted strings, etc.)

*But* when we enter a quoted string (which would be found inside an interpolation) we start parsing its content as if we were outside of an interpolation (it's the actual implementation: we re-start a parsing from scratch on the content of the quoted string -- after the un-escaping pass for escaped quotes). In that sense, the parsing of a quoted string inside an interpolation is the same as the parsing outside of interpolations.

Reply:
> is actually invalid because "'abc\\'" is the string with characters 'abc\' and is an invalid quoted string according to the grammar, since the end quote is escaped (=> it's missing another end quote).

I think I meant the characters `'abc\\'`. This is again getting confusing because of Python interpreting the backslashes as well. Is this one of the  cases where r strings are not working?

Thanks for clarifying the parsing logic.

Reply:
> I think I meant the characters `'abc\\'`. This is again getting confusing because of Python interpreting the backslashes as well. Is this one of the cases where r strings are not working?

Ok so the string with characters `'abc\\'` is the one from the same test:
```python
("str_quoted_trailing_esc_1", "'abc\\\\'", "abc\\"),
````

The reason why I didn't use r-strings is because the last string (`"abc\\"`) can't be converted into an r-string, and I found that using an r-string only for the first one could be confusing (when trying to compare before/after). It would look like this:
```python
("str_quoted_trailing_esc_1", r"'abc\\'", "abc\\"),
````
which I find a bit confusing because at first sight it looks like we have the same number of `\` in both strings.